R/animint.R

```diff
@@ -1022,69 +1037,89 @@ getCommonChunk <- function(built, chunk.vars, aes.list){
       built[, col.name] <- paste(built[, col.name])
     }
   }
-  built.by.chunk <- split(built, built[, chunk.vars], drop = TRUE)
-  if(length(built.by.chunk) == 1) return(NULL)
+
+  ## If there is only one chunk, then there is no point of making a
+  ## common data file.
+  chunk.rows.tab <- table(built[, chunk.vars])
+  if(length(chunk.rows.tab) == 1) return(NULL)
+
   ## If there is no group column, and all the chunks are the same
   ## size, then add one based on the row number.
   if(! "group" %in% names(built)){
-    chunk.rows.vec <- sapply(built.by.chunk, nrow)
-    chunk.rows <- chunk.rows.vec[1]
-    same.size <- chunk.rows == chunk.rows.vec
+    chunk.rows <- chunk.rows.tab[1]
+    same.size <- chunk.rows == chunk.rows.tab
+    order.args <- lapply(chunk.vars, function(order.col)built[[order.col]])
+    built <- built[do.call(order, order.args),]
     if(all(same.size)){
-      for(chunk.name in names(built.by.chunk)){
-        built.by.chunk[[chunk.name]]$group <- 1:chunk.rows
-      }
+      built$group <- 1:chunk.rows
     }else{
       ## do not save a common chunk file.
       return(NULL)
     }
   }
-  all.col.names <- names(built.by.chunk[[1]])
-  col.name.vec <- all.col.names[!all.col.names %in% chunk.vars]
-  values.by.group <- list()
-  for(chunk.name in names(built.by.chunk)){
-    chunk.df <- built.by.chunk[[chunk.name]]
-    chunk.by.group <- split(chunk.df, chunk.df$group, drop=TRUE)
-    for(group.name in names(chunk.by.group)){
-      values.by.group[[group.name]][[chunk.name]] <-
-        chunk.by.group[[group.name]]
-    }
-  }
-  is.common.mat <-
-    matrix(NA, length(values.by.group), length(col.name.vec),
-           dimnames=list(group=names(values.by.group),
-                         col.name=col.name.vec))
-  group.info.list <- list()
-  for(group.name in names(values.by.group)){
-    values.by.chunk <- values.by.group[[group.name]]
-    row.count.vec <- sapply(values.by.chunk, nrow)
-    same.size.chunks <- all(row.count.vec[1] == row.count.vec)
-    ## For every group, save values for creating common tsv later.
-    one.group.info <- values.by.chunk[[1]]
-    for(col.name in col.name.vec){
-      value.list <- lapply(values.by.chunk, function(df)df[[col.name]])
-      is.common.mat[group.name, col.name] <- if(same.size.chunks){
-        value.mat <- do.call(cbind, value.list)
-        least.missing <- which.min(colSums(is.na(value.mat)))
-        value.vec <- value.mat[, least.missing]
-        one.group.info[[col.name]] <- value.vec
-        all(value.vec == value.mat)
+
+  built.by.group <- split(built, built$group)
```

Comment:
this re-design of the common chunk computation is faster since it only has 1 level of for loops (through groups), as opposed to the previous code which had two levels (through groups and chunks)